pseudopotential_calculator/calculations/slab.py

```diff
@@ -26,36 +23,100 @@
     from matplotlib.lines import Line2D
 
 
-class SlabOptimizationParams(OptimizationParamsBase):
-    """Parameters of a slab calculation."""
+def _get_sorted_heights(atom: Atoms) -> np.ndarray[Any, np.dtype[np.float64]]:
+    heights = cast(np.ndarray[Any, np.dtype[np.float64]], atom.positions[:, 2])  # type: ignore bad lib
+    return np.unique(heights)[::-1]
 
-    @property
-    def kpoint_mp_grid(self: Self) -> str:
-        return f"{self.n_k_points} {self.n_k_points} {1}"
 
+def get_height_of_layer(atoms: Atoms, layer_idx: int) -> float:
+    heights = _get_sorted_heights(atoms)
+    return heights[layer_idx]
+
+
+def get_height_per_layer(atom: Atoms) -> float:
+    heights = _get_sorted_heights(atom)
+    return heights[-2]
+
+
+def get_n_slab_layer(atoms: Atoms) -> int:
+    heights = _get_sorted_heights(atoms)
+    return len(heights)
+
+
+def get_slab_width(atoms: Atoms) -> int:
+    # n_slab_atoms / n_atoms_per_repeat gives the number of slab units set to be repeated.
+    # take square root to get width since the repeat s set to be the same in x and y directions
+    n_slab_atoms = len(atoms.positions) - 1  # type: ignore bad lib
+    n_atoms_per_repeat = get_n_slab_layer(atoms) - 1
+    return np.sqrt(round(n_slab_atoms / n_atoms_per_repeat))
+
+
+def _get_n_vacuum_layers_from_slab(atoms: Atoms) -> float:
```

Comment:
I think there is a bug here??

Reply:
yeah the idx should be 0 for the first layer. It used to be 1 for the first layer but the logic is changed and I forgot to change this bit

Reply:
also I removed one get_height_per_layer function and put the correct one(should be, I tested it with some data) in to the function